src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -178,7 +178,7 @@ public void DisconnectFromWorkspace()
 
             private void EnqueueParseSnapshotTask(Document newDocument)
             {
-                if (newDocument != null)
+                if (newDocument != null && newDocument.SupportsSyntaxTree)
```

Comment:
1. why not just export this classifier provider for C#/VB?  Then this wouldn't ever even run for languages that don't support syntax trees?
2. i need to understand the issue more deeply.  We've have this type a *long* time.  It hasn't been a problem yet for F# or TS/JS.  So why now is there suddenly an issue with a crash here?

Reply:
I'm still trying to figure out what's the root cause is, but I think the crash is directly caused by VS providing us a buffer of content type "CSharp" for a xaml document.

Here's the stacktrace after the renaming of xaml file that eventually leads to the crash.
```
Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.Classification.SyntacticClassificationTaggerProvider.TagComputer.EnqueueParseSnapshotTask(Microsoft.CodeAnalysis.Document newDocument) Line 181
	at C:\github\roslyn\src\EditorFeatures\Core\Implementation\Classification\SyntacticClassificationTaggerProvider.TagComputer.cs(181)
Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.Classification.SyntacticClassificationTaggerProvider.TagComputer.ConnectToWorkspace(Microsoft.CodeAnalysis.Workspace workspace) Line 158
	at C:\github\roslyn\src\EditorFeatures\Core\Implementation\Classification\SyntacticClassificationTaggerProvider.TagComputer.cs(158)
Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.Classification.SyntacticClassificationTaggerProvider.TagComputer.TagComputer(Microsoft.VisualStudio.Text.ITextBuffer subjectBuffer, Microsoft.CodeAnalysis.Editor.IForegroundNotificationService notificationService, Microsoft.CodeAnalysis.Shared.TestHooks.IAsynchronousOperationListener asyncListener, Microsoft.CodeAnalysis.Editor.Shared.Utilities.ClassificationTypeMap typeMap, Microsoft.CodeAnalysis.Editor.Implementation.Classification.SyntacticClassificationTaggerProvider taggerProvider) Line 92
	at C:\github\roslyn\src\EditorFeatures\Core\Implementation\Classification\SyntacticClassificationTaggerProvider.TagComputer.cs(92)
Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.Classification.SyntacticClassificationTaggerProvider.CreateTagger<Microsoft.VisualStudio.Text.Tagging.IClassificationTag>(Microsoft.VisualStudio.Text.ITextBuffer buffer) Line 51
	at C:\github\roslyn\src\EditorFeatures\Core\Implementation\Classification\SyntacticClassificationTaggerProvider.cs(51)
Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.Tagging.Implementation.TagAggregator<Microsoft.VisualStudio.Text.Tagging.IClassificationTag>.InstantiateBuferTagger(System.Lazy<Microsoft.VisualStudio.Text.Tagging.ITaggerProvider, Microsoft.VisualStudio.Text.Tagging.INamedTaggerMetadata> factory, Microsoft.VisualStudio.Text.ITextBuffer textBuffer, System.Collections.Generic.IList<Microsoft.VisualStudio.Text.Tagging.ITagger<Microsoft.VisualStudio.Text.Tagging.IClassificationTag>> newTaggers)
Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.Tagging.Implementation.TagAggregator<Microsoft.VisualStudio.Text.Tagging.IClassificationTag>.AddTaggers(Microsoft.VisualStudio.Text.ITextBuffer textBuffer, System.Collections.Generic.IList<Microsoft.VisualStudio.Text.Tagging.ITagger<Microsoft.VisualStudio.Text.Tagging.IClassificationTag>> newTaggers)
Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.Tagging.Implementation.TagAggregator<Microsoft.VisualStudio.Text.Tagging.IClassificationTag>.OnContentTypeChanged(object sender, Microsoft.VisualStudio.Text.ContentTypeChangedEventArgs e)
Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent<Microsoft.VisualStudio.Text.ContentTypeChangedEventArgs>(object sender, System.EventHandler<Microsoft.VisualStudio.Text.ContentTypeChangedEventArgs> eventHandlers, Microsoft.VisualStudio.Text.ContentTypeChangedEventArgs args)
Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.Implementation.BaseBuffer.ContentTypeChangedEventRaiser.RaiseEvent(Microsoft.VisualStudio.Text.Implementation.BaseBuffer baseBuffer, bool immediate)
Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.Implementation.BufferGroup.RaiseEvents()
Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.Implementation.BufferGroup.FinishEdit()
Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.Implementation.BaseBuffer.TextBufferEdit.Apply()
Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.Implementation.BaseBuffer.ChangeContentType(Microsoft.VisualStudio.Utilities.IContentType newContentType, object editTag)
Microsoft.VisualStudio.Editor.Implementation.dll!Microsoft.VisualStudio.Editor.Implementation.TextDocData.SetLanguageServiceID(ref System.Guid guidLangService, bool changeContentType, bool sendChangedEvent)
Microsoft.VisualStudio.Editor.Implementation.dll!Microsoft.VisualStudio.Editor.Implementation.TextDocData.ChooseLanguageServicePerBufferMoniker(bool shouldSendEvent)
Microsoft.VisualStudio.Editor.Implementation.dll!Microsoft.VisualStudio.Editor.Implementation.TextDocData.SetData_Moniker(string newMoniker, bool shouldSendEvent)
Microsoft.VisualStudio.Editor.Implementation.dll!Microsoft.VisualStudio.Editor.Implementation.TextDocData.SetData(ref System.Guid riidKey, object vtData)
Microsoft.VisualStudio.Editor.Implementation.dll!Microsoft.VisualStudio.Editor.Implementation.TextDocData.RenameDocData(uint grfAttribs, Microsoft.VisualStudio.Shell.Interop.IVsHierarchy pHierNew, uint itemidNew, string pszMkDocumentNew)
[Native to Managed Transition]
[Managed to Native Transition]
Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.Flavor.FlavoredProjectBase.SetProperty(uint itemId, int propId, object property)
Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.Flavor.FlavoredProjectBase.Microsoft.VisualStudio.Shell.Interop.IVsHierarchy.SetProperty(uint itemId, int propId, object property)
[Native to Managed Transition]
[Managed to Native Transition]
Microsoft.VisualStudio.Shell.TreeNavigation.HierarchyProvider.dll!Microsoft.VisualStudio.PlatformUI.HierarchyItem.RenameLabel.set(string value)
Microsoft.VisualStudio.Shell.TreeNavigation.HierarchyProvider.dll!Microsoft.VisualStudio.PlatformUI.HierarchyRenameItemTransaction.Commit(Microsoft.Internal.VisualStudio.PlatformUI.RenameItemCompletionFocusBehavior completionFocusBehavior)
Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.EditableItemTextBox.OnKeyDown(System.Windows.Input.KeyEventArgs e)
PresentationCore.dll!System.Windows.UIElement.OnKeyDownThunk(object sender, System.Windows.Input.KeyEventArgs e)
PresentationCore.dll!System.Windows.Input.KeyEventArgs.InvokeEventHandler(System.Delegate genericHandler, object genericTarget)
PresentationCore.dll!System.Windows.RoutedEventArgs.InvokeHandler(System.Delegate handler, object target)
PresentationCore.dll!System.Windows.RoutedEventHandlerInfo.InvokeHandler(object target, System.Windows.RoutedEventArgs routedEventArgs)
PresentationCore.dll!System.Windows.EventRoute.InvokeHandlersImpl(object source, System.Windows.RoutedEventArgs args, bool reRaised)
PresentationCore.dll!System.Windows.UIElement.RaiseEventImpl(System.Windows.DependencyObject sender, System.Windows.RoutedEventArgs args)
PresentationCore.dll!System.Windows.UIElement.RaiseTrustedEvent(System.Windows.RoutedEventArgs args)
PresentationCore.dll!System.Windows.UIElement.RaiseEvent(System.Windows.RoutedEventArgs args, bool trusted)
PresentationCore.dll!System.Windows.Input.InputManager.ProcessStagingArea()
PresentationCore.dll!System.Windows.Input.InputManager.ProcessInput(System.Windows.Input.InputEventArgs input)
PresentationCore.dll!System.Windows.Input.InputProviderSite.ReportInput(System.Windows.Input.InputReport inputReport)
PresentationCore.dll!System.Windows.Interop.HwndKeyboardInputProvider.ReportInput(System.IntPtr hwnd, System.Windows.Input.InputMode mode, int timestamp, System.Windows.Input.RawKeyboardActions actions, int scanCode, bool isExtendedKey, bool isSystemKey, int virtualKey)
PresentationCore.dll!System.Windows.Interop.HwndKeyboardInputProvider.ProcessKeyAction(ref System.Windows.Interop.MSG msg, ref bool handled)
PresentationCore.dll!System.Windows.Interop.HwndSource.CriticalTranslateAccelerator(ref System.Windows.Interop.MSG msg, System.Windows.Input.ModifierKeys modifiers)
PresentationCore.dll!System.Windows.Interop.HwndSource.OnPreprocessMessage(object param)
WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate callback, object args, int numArgs)
WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen(object source, System.Delegate callback, object args, int numArgs, System.Delegate catchHandler)
WindowsBase.dll!System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority priority, System.TimeSpan timeout, System.Delegate method, object args, int numArgs)
WindowsBase.dll!System.Windows.Threading.Dispatcher.Invoke(System.Windows.Threading.DispatcherPriority priority, System.Delegate method, object arg)
PresentationCore.dll!System.Windows.Interop.HwndSource.OnPreprocessMessageThunk(ref System.Windows.Interop.MSG msg, ref bool handled)
PresentationCore.dll!System.Windows.Interop.HwndSource.WeakEventPreprocessMessage.OnPreprocessMessage(ref System.Windows.Interop.MSG msg, ref bool handled)
WindowsBase.dll!System.Windows.Interop.ComponentDispatcherThread.RaiseThreadMessage(ref System.Windows.Interop.MSG msg)
Microsoft.VisualStudio.Platform.WindowManagement.dll!Microsoft.VisualStudio.Platform.WindowManagement.WindowManagerService.PreTranslateInput(ref Microsoft.VisualStudio.OLE.Interop.MSG oleMsg, bool handled)

```

Reply:
Also I just tried renaming a f# file to .cs in solution explorer, it indeed triggers the same crash. I suspect the same for documents in TypeScript project

Reply:
@CyrusNajmabadi I have added some comments

Reply:
problem is content type is editor concept and document is our concept. so, not sure how we coordinate these two when content type get changed.

Reply:
one thing we could try would be
changing this "GetDocumentIdInCurrentContext".

most of time, this is the API we convert buffer to document. right now, it gets our buffer to document where we get our buffer from editor buffer through AsTextContainer().

we could add new API "GetDocumentIdInCurrentContext" that accepts editor buffer directly and in the API, it checks content type of the buffer to be same as language of the document we get out. if they are different, return null.